## 1. Introduction to Confidence Intervals for Proportions

A **confidence interval** provides a range of values within which we expect the true population parameter (in this case, the proportion) to lie, given a certain level of confidence (e.g., 95%). For population proportions, two commonly used methods to calculate these intervals are:

- **Wald Interval**: Based on the normal approximation to the binomial distribution.
- **Wilson Interval**: Addresses some limitations of the Wald interval, especially with small sample sizes or proportions near 0 or 1.

## 2. Wald Confidence Interval

The Wald interval is calculated using the formula:

$$ \hat{p} \pm z_{\alpha/2} \sqrt{\frac{\hat{p}(1 - \hat{p})}{n}} $$

where:
- $\hat{p}$ is the sample proportion.
- $n$ is the sample size.
- $z_{\alpha/2}$ is the critical value from the standard normal distribution corresponding to the desired confidence level.

However, the Wald interval can perform poorly, especially with small sample sizes or when the true proportion is near 0 or 1. In such cases, the interval may extend beyond the [0, 1] range, which is not meaningful for proportions.

## 3. Wilson (Score) Confidence Interval

The Wilson interval is given by:

$$ \frac{\hat{p} + \frac{z^2}{2n} \pm z \sqrt{\frac{\hat{p}(1 - \hat{p})}{n} + \frac{z^2}{4n^2}}}{1 + \frac{z^2}{n}} $$

The Wilson interval is asymmetric and remains within the [0, 1] range, providing more realistic estimates for proportions. It has better coverage properties, meaning the actual confidence level is closer to the nominal level, even with small sample sizes or extreme proportions.

## 4. Implementation in Python

Let's implement both intervals in Python and compare them with a small sample size where the sample proportion is small.

In [1]:
import numpy as np
from scipy.stats import norm

# Sample data
n = 30  # Sample size
x = 1  # Number of successes
p_hat = x / n  # Sample proportion

# Confidence level
conf_level = 0.95
z = norm.ppf(1 - (1 - conf_level) / 2)

# Wald confidence interval
wald_se = np.sqrt(p_hat * (1 - p_hat) / n)
wald_lower = p_hat - z * wald_se
wald_upper = p_hat + z * wald_se

# Wilson confidence interval
denominator = 1 + z**2 / n
centre_adj = p_hat + z**2 / (2 * n)
adj_std_dev = np.sqrt((p_hat * (1 - p_hat) + z**2 / (4 * n)) / n)
wilson_lower = (centre_adj - z * adj_std_dev) / denominator
wilson_upper = (centre_adj + z * adj_std_dev) / denominator

# Display results
print(f"Sample proportion (p̂): {p_hat:.4f}")
print(f"Wald CI:    ({wald_lower:.4f}, {wald_upper:.4f})")
print(f"Wilson CI:  ({wilson_lower:.4f}, {wilson_upper:.4f})")

Sample proportion (p̂): 0.0333
Wald CI:    (-0.0309, 0.0976)
Wilson CI:  (0.0059, 0.1667)


## 5. Interpretation of Results

In this example, with a sample size of 30 and only 1 success ($\hat{p} \approx 0.0333$), the Wald confidence interval produces a negative lower bound, which is not meaningful for a proportion that must lie between 0 and 1. Conversely, the Wilson interval provides a more appropriate range, staying within the valid bounds for a proportion.

This example illustrates the limitations of the Wald interval in cases with small sample proportions and highlights the robustness of the Wilson interval in providing accurate confidence intervals under such conditions.